File: src/claw_mem/values/feedback_handler.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Feedback Handler - Feedback processing mechanism
"""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, List, Optional

from claw_mem.values import UserValueStore


class FeedbackStatus(Enum):
    """Feedback status"""

    PENDING = "pending"  # Pending
    ACCEPTED = "accepted"  # Accepted
    REJECTED = "rejected"  # Rejected
    EXPIRED = "expired"  # Expired


@dataclass
class ValueSuggestion:
    """Value suggestion"""

    id: str
    user_id: str
    suggestion_type: str  # "principle", "preference", "red_line"
    content: str
    evidence: List[str] = field(default_factory=list)
    status: FeedbackStatus = FeedbackStatus.PENDING
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    responded_at: Optional[datetime] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "user_id": self.user_id,
            "suggestion_type": self.suggestion_type,
            "content": self.content,
            "evidence": self.evidence,
            "status": self.status.value,
            "created_at": self.created_at.isoformat(),
            "responded_at": self.responded_at.isoformat() if self.responded_at else None,
        }


class FeedbackHandler:
    """Feedback processor - Manages user confirmation of values"""

    def __init__(self, value_store: Optional[UserValueStore] = None):
        """Initialize feedback processor

        Args:
            value_store: User values storage
        """
        self.value_store = value_store or UserValueStore()

        # Pending suggestions
        self._pending_suggestions: Dict[str, List[ValueSuggestion]] = {}

        # Suggestion history
        self._suggestion_history: List[ValueSuggestion] = []

    def request_confirmation(
        self, user_id: str, value_type: str, content: str, evidence: List[str] = None
    ) -> ValueSuggestion:
        """Request user confirmation of values

        Args:
            user_id: User ID
            value_type: Value type ("principle", "preference", "red_line")
            content: Value content
            evidence: Evidence list

        Returns:
            ValueSuggestion: Created suggestion
        """
        import uuid

        suggestion = ValueSuggestion(
            id=str(uuid.uuid4())[:8],
            user_id=user_id,
            suggestion_type=value_type,
            content=content,
            evidence=evidence or [],
        )

        # Add to pending list
        if user_id not in self._pending_suggestions:
            self._pending_suggestions[user_id] = []

        self._pending_suggestions[user_id].append(suggestion)
        self._suggestion_history.append(suggestion)

        return suggestion

    def process_feedback(self, suggestion_id: str, accepted: bool) -> bool:
        """Process user feedback

        Args:
            suggestion_id: Suggestion ID
            accepted: Whether accepted

        Returns:
            bool: Whether successfully processed
        """
        # Find suggestion
        suggestion = None
        for s in self._suggestion_history:
            if s.id == suggestion_id:
                suggestion = s
                break

        if not suggestion:
            return False

        # Update status
        suggestion.status = FeedbackStatus.ACCEPTED if accepted else FeedbackStatus.REJECTED
        suggestion.responded_at = datetime.now(timezone.utc)

        # If accepted, update to value store
        if accepted:
            user_id = suggestion.user_id

            if suggestion.suggestion_type == "principle":
                self.value_store.save_principle(user_id, suggestion.content)

            elif suggestion.suggestion_type == "preference":
                # Preference needs key-value parsing
                # Simplified: assume content format is "key:value"
                if ":" in suggestion.content:
                    key, value = suggestion.content.split(":", 1)
                    self.value_store.save_preference(user_id, key.strip(), value.strip())

            elif suggestion.suggestion_type == "red_line":
                self.value_store.save_red_line(user_id, suggestion.content)

        # Remove from pending list
        user_id = suggestion.user_id
        if user_id in self._pending_suggestions:
            self._pending_suggestions[user_id] = [
                s for s in self._pending_suggestions[user_id] if s.id != suggestion_id
            ]

        return True

    def suggest_update(self, suggestion: Dict[str, Any]) -> ValueSuggestion:
        """Suggest updating values

        Args:
            suggestion: Suggestion data

        Returns:
            ValueSuggestion: Created suggestion
        """
        return self.request_confirmation(
            user_id=suggestion["user_id"],
            value_type=suggestion.get("type", "principle"),
            content=suggestion["content"],
            evidence=suggestion.get("evidence", []),
        )

    def get_pending_suggestions(self, user_id: str) -> List[ValueSuggestion]:
        """Get pending suggestions

        Args:
            user_id: User ID

        Returns:
            List[ValueSuggestion]: List of pending suggestions
        """
        return self._pending_suggestions.get(user_id, [])

    def get_accepted_suggestions(self, user_id: str) -> List[ValueSuggestion]:
        """Get accepted suggestions

        Args:
            user_id: User ID

        Returns:
            List[ValueSuggestion]: List of accepted suggestions
        """
        return [
            s
            for s in self._suggestion_history
            if s.user_id == user_id and s.status == FeedbackStatus.ACCEPTED
        ]

    def get_rejected_suggestions(self, user_id: str) -> List[ValueSuggestion]:
        """Get rejected suggestions

        Args:
            user_id: User ID

        Returns:
            List[ValueSuggestion]: List of rejected suggestions
        """
        return [
            s
            for s in self._suggestion_history
            if s.user_id == user_id and s.status == FeedbackStatus.REJECTED
        ]

    def clear_expired(self, max_age_hours: int = 24) -> int:
        """Clear expired suggestions

        Args:
            max_age_hours: Max retention time (hours)

        Returns:
            int: Number of cleared entries
        """
        now = datetime.now(timezone.utc)
        expired = []

        for suggestion in self._suggestion_history:
            if suggestion.status == FeedbackStatus.PENDING:
                age = (now - suggestion.created_at).total_seconds() / 3600
                if age > max_age_hours:
                    suggestion.status = FeedbackStatus.EXPIRED
                    suggestion.responded_at = now
                    expired.append(suggestion)

        return len(expired)


__all__ = [
    "FeedbackStatus",
    "ValueSuggestion",
    "FeedbackHandler",
]

File: src/values/feedback_handler.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * Feedback Handler - Feedback processing mechanism
 */

import * as crypto from "crypto";
import { UserValueStore } from "./user_value_store.js";

// ── Enums & types ──────────────────────────────────────────────────────

export enum FeedbackStatus {
  PENDING = "pending",
  ACCEPTED = "accepted",
  REJECTED = "rejected",
  EXPIRED = "expired",
}

export interface ValueSuggestion {
  id: string;
  userId: string;
  suggestionType: string; // "principle" | "preference" | "red_line"
  content: string;
  evidence: string[];
  status: FeedbackStatus;
  createdAt: Date;
  respondedAt: Date | null;
}

export function valueSuggestionToDict(s: ValueSuggestion): Record<string, unknown> {
  return {
    id: s.id,
    user_id: s.userId,
    suggestion_type: s.suggestionType,
    content: s.content,
    evidence: s.evidence,
    status: s.status,
    created_at: s.createdAt.toISOString(),
    responded_at: s.respondedAt ? s.respondedAt.toISOString() : null,
  };
}

// ── Handler ────────────────────────────────────────────────────────────

export class FeedbackHandler {
  private _pendingSuggestions: Record<string, ValueSuggestion[]> = {};
  private _suggestionHistory: ValueSuggestion[] = [];

  constructor(private _valueStore: UserValueStore = new UserValueStore()) {}

  /** Request user confirmation of values. */
  requestConfirmation(
    userId: string,
    valueType: string,
    content: string,
    evidence: string[] = [],
  ): ValueSuggestion {
    const suggestion: ValueSuggestion = {
      id: crypto.randomUUID().slice(0, 8),
      userId,
      suggestionType: valueType,
      content,
      evidence,
      status: FeedbackStatus.PENDING,
      createdAt: new Date(),
      respondedAt: null,
    };

    // Add to pending list
    if (!this._pendingSuggestions[userId]) {
      this._pendingSuggestions[userId] = [];
    }
    this._pendingSuggestions[userId].push(suggestion);
    this._suggestionHistory.push(suggestion);

    return suggestion;
  }

  /** Process user feedback. Returns true if successfully processed. */
  processFeedback(suggestionId: string, accepted: boolean): boolean {
    // Find suggestion
    const suggestion = this._suggestionHistory.find((s) => s.id === suggestionId);
    if (!suggestion) return false;

    // Update status
    suggestion.status = accepted ? FeedbackStatus.ACCEPTED : FeedbackStatus.REJECTED;
    suggestion.respondedAt = new Date();

    // If accepted, update to value store
    if (accepted) {
      const userId = suggestion.userId;

      if (suggestion.suggestionType === "principle") {
        this._valueStore.savePrinciple(userId, suggestion.content);
      } else if (suggestion.suggestionType === "preference") {
        // Preference needs key-value parsing
        // Simplified: assume content format is "key:value"
        const colonIdx = suggestion.content.indexOf(":");
        if (colonIdx !== -1) {
          const key = suggestion.content.slice(0, colonIdx).trim();
          const value = suggestion.content.slice(colonIdx + 1).trim();
          this._valueStore.savePreference(userId, key, value);
        }
      } else if (suggestion.suggestionType === "red_line") {
        this._valueStore.saveRedLine(userId, suggestion.content);
      }
    }

    // Remove from pending list
    const userId = suggestion.userId;
    if (this._pendingSuggestions[userId]) {
      this._pendingSuggestions[userId] = this._pendingSuggestions[userId].filter(
        (s) => s.id !== suggestionId,
      );
    }

    return true;
  }

  /** Suggest updating values. */
  suggestUpdate(suggestion: Record<string, unknown>): ValueSuggestion {
    return this.requestConfirmation(
      suggestion.user_id as string,
      (suggestion.type as string) || "principle",
      suggestion.content as string,
      (suggestion.evidence as string[]) || [],
    );
  }

  /** Get pending suggestions for a user. */
  getPendingSuggestions(userId: string): ValueSuggestion[] {
    return this._pendingSuggestions[userId] || [];
  }

  /** Get accepted suggestions for a user. */
  getAcceptedSuggestions(userId: string): ValueSuggestion[] {
    return this._suggestionHistory.filter(
      (s) => s.userId === userId && s.status === FeedbackStatus.ACCEPTED,
    );
  }

  /** Get rejected suggestions for a user. */
  getRejectedSuggestions(userId: string): ValueSuggestion[] {
    return this._suggestionHistory.filter(
      (s) => s.userId === userId && s.status === FeedbackStatus.REJECTED,
    );
  }

  /** Clear expired suggestions. Returns count of cleared entries. */
  clearExpired(maxAgeHours: number = 24): number {
    const now = new Date();
    let expired = 0;

    for (const suggestion of this._suggestionHistory) {
      if (suggestion.status === FeedbackStatus.PENDING) {
        const ageMs = now.getTime() - suggestion.createdAt.getTime();
        const ageHours = ageMs / (1000 * 60 * 60);
        if (ageHours > maxAgeHours) {
          suggestion.status = FeedbackStatus.EXPIRED;
          suggestion.respondedAt = now;
          expired++;
        }
      }
    }

    return expired;
  }
}
